File: ios/Classes/BannerView/BannerViewEventListener.swift
//
//  BannerViewEventListener.swift
//  clever_ads_solutions
//
//  Created by MacMini on 2.04.24.
//

import Foundation
import CleverAdsSolutions
import Flutter

class BannerViewEventListener: NSObject, CASBannerDelegate, FlutterStreamHandler {
    var flutterIds: [Int:String] = [Int:String]()
    private var eventSink: FlutterEventSink?
    
    func bannerAdViewDidLoad(_ view: CASBannerView) {
        if let sink = self.eventSink {
            sink([
                "id": flutterIds[view.tag],
                "event": "onAdViewLoaded"
            ])
        }
    }
    
    func bannerAdView(_ adView: CASBannerView, willPresent impression: CASImpression) {
        if let sink = self.eventSink {
            sink([
                "id": flutterIds[adView.tag],
                "event": "onAdViewPresented",
                "data": impression.toDict()
            ])
        }
    }
    
    func bannerAdViewDidRecordClick(_ adView: CASBannerView) {
        if let sink = self.eventSink {
            sink([
                "id": flutterIds[adView.tag],
                "event": "onAdViewClicked"
            ])
        }
    }
    
    func bannerAdView(_ adView: CASBannerView, didFailWith error: CASError) {
        if let sink = self.eventSink {
            sink([
                "id": flutterIds[adView.tag],
                "event": "onAdViewFailed",
                "data": error.message
            ])
        }
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bannerview/BannerViewEventListener.kt
package com.cleveradssolutions.plugin.flutter.bannerview

import android.util.Log
import com.cleveradssolutions.plugin.flutter.util.toMap
import com.cleversolutions.ads.AdError
import com.cleversolutions.ads.AdImpression
import com.cleversolutions.ads.AdViewListener
import com.cleversolutions.ads.android.CASBannerView
import io.flutter.plugin.common.EventChannel

class BannerViewEventListener : AdViewListener, EventChannel.StreamHandler {

    val flutterIds = HashMap<Int, String>()
    private var eventSink: EventChannel.EventSink? = null

    override fun onAdViewLoaded(view: CASBannerView) {
        eventSink?.success(
            mapOf(
                "id" to flutterIds[view.id],
                "event" to "onAdViewLoaded"
            )
        )
    }

    override fun onAdViewFailed(view: CASBannerView, error: AdError) {
        eventSink?.success(
            mapOf(
                "id" to flutterIds[view.id],
                "event" to "onAdViewFailed",
                "data" to error.message
            )
        )
    }

    override fun onAdViewPresented(view: CASBannerView, info: AdImpression) {
        eventSink?.success(
            mapOf(
                "id" to flutterIds[view.id],
                "event" to "onAdViewPresented",
                "data" to info.toMap()
            )
        )
    }

    override fun onAdViewClicked(view: CASBannerView) {
        eventSink?.success(
            mapOf(
                "id" to flutterIds[view.id],
                "event" to "onAdViewClicked"
            )
        )
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}